#7995 Micro-ROS configuration NuttX tutorial (#140)
micro-ROS/micro-ROS.github.io: _docs/tutorials/advanced/nuttx/add_microros_config/index.md

    - STM32 MCU.
    - RAM: 125 Kb.
    - Flash: 512 kb.
    - Serial communication peripheral.

- USB-TTY serial cable.

## Workspace set-up

We're going to create the Micro-ROS workspace, to do so, we will execute the following commands on a console:

```
source /opt/ros/$ROS_DISTRO/setup.bash

mkdir uros_ws && cd uros_ws

git clone -b $ROS_DISTRO https://github.com/micro-ROS/micro-ros-build.git src/micro-ros-build

rosdep update && rosdep install --from-path src --ignore-src -y

colcon build

source install/local_setup.bash
```

## Set the base configuration

 In this guide, for example, we will give Micro-ROS support for the [Olimex-STM32-H407 board](https://www.olimex.com/Products/ARM/ST/STM32-H407/open-source-hardware).

This board is a simplified version of our supported Olimex-STM32-E407 board. 

On the previous console, execute the following commands to set the basic configuration:
```bash
ros2 run micro_ros_setup create_firmware_ws.sh nuttx olimex-stm32-h407
ros2 run micro_ros_setup configure_firmware.sh nsh
```

If everything goes fine, it should output the next message:
```bash
Copy files
Refreshing...
```

> Now is set the basic NSH configuration. On the forward steps, we will set the required configuration to run Micro-ROS on this board.

Reviewer: ```suggestion
Now the basic NSH configuration is set. With the steps that follow, we will set the required configuration to run Micro-ROS on this board.
```

## Micro-ROS configuration.

> On the previous console execute the next commands:

Reviewer: ```suggestion
In the same console as before, run the commands:
```

```bash
cd firmware/NuttX
make menuconfig
```

> This should show this menu:

Reviewer: ```suggestion
The menu below should appear:
```

![](images/nuttx_menuconfig.png)

> Note: Remember that you should follow this previous tutorial before starting it. [Adding Micro-ROS to a NuttX board configuration](https://micro-ros.github.io/docs/tutorials/advanced/nuttx/microros_nuttx_bsp/)

Reviewer: You already said it before. Are you sure it makes sense to repeat it here? Maybe the best thin to do is to remove it from here, and also from above, and create instead a 'Disclaimer' in bold above in a new line, close to where the other Disclaimer is. (I prefer the format where the disclaimers are not section but something inline like "Disclaimer!")

Author: Done!

> On the menu, you need to set the next configuration.

Reviewer: ```suggestion
In the menu, you need to set the following configuration.
```

- System Type > STM32 Peripheral Support > USART3
- RTOS Features > Clocks and Timers > Support CLOCK_MONOTONIC
- Device Drivers > Serial Driver Support > Serial TERMIOS support
- Library Routines > Standard Math Library
- Library Routines > sizeof(_Bool) is a 8-bits
- Library Routines > Build uClibc++ (must be installed) 
- Application Configuration > micro-ROS
- Application Configuration > micro-ROS > Transport > Serial Transport
- Application Configuration > Examples > microROS Publisher

Push two times the key "ESC" and set yes, when it asks you if you want to save it.

> Now the configuration is properly set, is just left to compile it. To do so, continue on the same console and execute the commands:

Reviewer: ```suggestion
Now the configuration is properly set, the only thing left is to compile it. To do so, continue in the same console and execute the commands:
```

```
cd ../..
ros2 run micro_ros_setup build_firmware.sh
```

> If everything goes fine, it should return the next output:

Reviewer: ```suggestion
If everything goes fine, it should return the following output:
```
